webapi/src/main/scala/org/knora/webapi/slice/infrastructure/Scope.scala

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright © 2021 - 2024 Swiss National Data and Service Center for the Humanities and/or DaSCH Service Platform contributors.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.knora.webapi.slice.infrastructure
+
+import org.knora.webapi.slice.admin.domain.model.KnoraProject.Shortcode
+import org.knora.webapi.slice.infrastructure.ScopeValue.Admin
+
+final case class Scope(values: Set[ScopeValue]) { self =>
+
+  def toScopeString: String = self.values.map(_.toScopeString).mkString(" ")
+
+  def +(addThis: ScopeValue): Scope =
+    if (self.values.contains(Admin) || addThis == Admin) {
+      Scope(Set(Admin))
+    } else {
+      values.find(_.merge(addThis).size == 1) match {
+        case Some(value) => Scope(self.values - value ++ value.merge(addThis))
+        case None        => Scope(self.values + addThis)
+      }
+    }
+}
+
+object Scope {
+  val empty: Scope = Scope(Set.empty)
+  val admin: Scope = Scope(Set(ScopeValue.Admin))
+
+  def from(scopeValues: Seq[ScopeValue]): Scope = scopeValues.foldLeft(Scope.empty)(_ + _)
+}
+
+sealed trait ScopeValue {
+  def toScopeString: String
+  final def merge(other: ScopeValue): Set[ScopeValue] = ScopeValue.merge(this, other)
+}
+
+object ScopeValue {
+  final case class Read(project: Shortcode) extends ScopeValue {
+    override def toScopeString: String = s"read:project:${project.value}"
+  }
+
+  final case class Write(project: Shortcode) extends ScopeValue {
+    override def toScopeString: String = s"write:project:${project.value}"
+  }
+
+  final case object Admin extends ScopeValue {
+    override def toScopeString: String = "admin"
+  }
+
+  def merge(one: ScopeValue, two: ScopeValue): Set[ScopeValue] =
+    (one, two) match {
+      case (Admin, _) | (_, Admin)            => Set(Admin)
+      case (Write(p1), Write(p2)) if p1 == p2 => Set(Write(p1))
+      case (Write(p1), Read(p2)) if p1 == p2  => Set(Write(p1))
+      case (Read(p1), Write(p2)) if p1 == p2  => Set(Write(p1))
+      case (Read(p1), Read(p2)) if p1 == p2   => Set(Read(p1))
+      case (a, b)                             => Set(a, b)
+    }
```

Comment:
This finds the "biggest" items of Set(one, two) by partial ordering Admin > Write > Read, with Write/Reads being not comparable if their contents differ, but there's not stdlib support, I'm afraid.

Reply:
Reason why I'm thinking about this is, if you would phrase it this way, then you wouldn't have the have the second filtering in `def +(addThis)`... Nevertheless, it looks to me like you don't need it anyway, let me consult your tests.

Reply:
fix pushed

Reply:
Thanks